### ISSUE TITLE
do not use `IndexKeyValueToPartialRecord` in the Cascades planner

### ISSUE THREAD

**Issue by normen662** [FoundationDB/fdb-record-layer#2907]
This issue is to exclusively use `IndexKeyValueToPartialRecord` in the heuristic planner and use a `RecordConstructorValue` or similar to compute the partial record that the `RecordQueryCoveringIndexScanPlan` can then return.

**Comment by hatyo** [comment]
Having an RCV-based interpreter of the `IndexEntry` makes sense, but it creates a problem for covering index scans, because `RecordConstructorValue` refuses to leave a non-nullable field unset: a column evaluating to null is only skipped once the field's type is verified nullable.
Take table `T(a, b, c, d)` and a value index `I` on keys `(c, d)`, no value part. For `select c, d from t where c = 20 and d < 40` the planner picks `I` and builds a covering scan. The entry is `KEY(c, d, <record type>, <primary key>)` with an empty value part, so interpreting it against the base record type `Record(a, b, c, d)` leaves `a` and `b` with null columns, and evaluation throws.
Today's copier-based reader gets away with this because a partial record simply omits the columns the index doesn't contain, the problem is, this particular assumption is *never* stated by the operator's type. The RCV is the first thing to refuse that assumption, which makes it visible: the operator promises the full record type while only ever delivering the indexed columns and the primary key.

**Comment by normen662** [comment]
The old `IndexKeyValueToPartialRecord` omits the fields that are not in the index as stated. I think we do a check somewhere to make sure that those fields are `optional` and never `required`; otherwise, I think protobuf may throw as well; I cannot remember if that check is in FRL or in protobuf. So yes, if the RCV encounters a non-nullable field, it must set it and I think that is the same as it is right now with `IndexKeyValueToPartialRecord`.

In other words, creating a `RCV(null, "someString", 15)` over a schema `(a: BIGINT, b: STRING, C: INT)` will only ever work if `a` is nullable and that's the same as before.

Somewhat related: In relational do we make non-nullable fields required on proto-level?

**Comment by hatyo** [comment]
Yes, I don't think this is related to nullability, it is much related to accepting the absence of a field from a protobuf message, which is ok as long as the field is marked as `optional`.

OTOH, the `RecordConstructorValue` is allowing the field to be `null` if and only if its type is nullable, regardless to whether the underlying descriptor is marking the field as `optional` or not.

> Somewhat related: In relational do we make non-nullable fields required on proto-level?

FRL does not support not-nullable types, except for not-nullable array elements, which is `repeated` (not `required`, and not `optional`), here is another example:

```sql
create table t3(id bigint, arr bigint array not null, c bigint, d bigint, primary key(id))
create index mv3 as select c, d from t3 order by c, d
```

for a query like:
```sql
select c, d from t3 where c = 20
```

In the current implementation, the copier seems to be looking at the nullability / `required` label of the field it is currently copying only (the covered fields):

https://github.com/FoundationDB/fdb-record-layer/blob/80db56e8446f09125b19c35ffeb38aedd8af0fb7/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/IndexKeyValueToPartialRecord.java#L457-L463

and so it will happily create a dynamic message with `{c, d}` (and primary key) only, completely ditching `arr`.

in a RCV-based implementation, that `RecordConstructorValue` will fail with a verify exception, because:
1. It is defined on the entire type, not just the covering columns, so it includes `arr`.
2. `arr` is marked as `not nullable`:

https://github.com/FoundationDB/fdb-record-layer/blob/80db56e8446f09125b19c35ffeb38aedd8af0fb7/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RecordConstructorValue.java#L129

I hope you see the issue now.
